File: src/Java/BehavioralPatterns/Mediator/CommunicationHub/src/Mediators/SensorCommandMediator.java
package Mediators;

import Participants.Participant;
import Participants.Sensor;
import Participants.User;

import java.util.List;
import java.util.stream.Collectors;

public class SensorCommandMediator extends GroupMediator {
    @Override
    public void notify(Participant sender, Object senderArgs) {
        List<Group> groupsOfParticipant = getGroups().stream()
                .filter(group -> group.doesParticipantExist(sender))
                .collect(Collectors.toList());

        List<Participant> receivers = groupsOfParticipant.stream()
                .flatMap(group -> group.getParticipants().stream())
                .filter(p -> p != sender)
                .distinct()
                .collect(Collectors.toList());

        if (senderArgs.equals("measure")) {
            receivers = receivers.stream().filter(r -> !(r instanceof User)).collect(Collectors.toList());
        } else {
            receivers = receivers.stream().filter(r -> !(r instanceof Sensor)).collect(Collectors.toList());
        }

        receivers.forEach(r -> r.receive(sender, senderArgs));
    }
}

File: src/JavaScript/BehavioralPatterns/Mediator/CommunicationHub/mediators/sensorCommandMediator.js

const GroupMediator = require("./groupMediator");

class SensorCommandMediator extends GroupMediator {
  notify(sender, senderArgs) {
    const groupsOfParticipant = this.groups.filter(group => group.doesParticipantExist(sender));

    let receivers = groupsOfParticipant
      .flatMap(group => group.participants)
      .filter(participant => participant !== sender)
      .filter((value, index, self) => self.indexOf(value) === index); // Remove duplicates

    if (senderArgs === "measure") {
      receivers = receivers.filter(receiver => !(receiver instanceof User));
    } else {
      receivers = receivers.filter(receiver => !(receiver instanceof Sensor));
    }

    receivers.forEach(receiver => receiver.receive(sender, senderArgs));
  }
}

module.exports = SensorCommandMediator;